pyatv/mrp/protocol.py

```diff
@@ -166,8 +166,8 @@ def message_received(self, message, _):
                 self._outstanding[identifier].semaphore, message)
             self._outstanding[identifier] = outstanding
             self._outstanding[identifier].semaphore.release()
-        else:
-            self._dispatch(message)
+
+        self._dispatch(message)
```

Comment:
@postlund, you are right, because of the initial DeviceInfoMessage I've changed the `self._dispatch(message)`.

Can you give an example please how can I store the message in protocol.py?

Reply:
The response (`DeviceInfoMessage`) is returned by `send_and_receive` here:

https://github.com/postlund/pyatv/blob/070946fd05d9c84bbfa2f3f0c98af62a459ec9e3/pyatv/mrp/protocol.py#L68

So you can just save it like (also add `self.device_info = None` in constructor to make pylint happy):

```python
    self.device_info = await self.send_and_receive(msg)
```

It is then available via `protocol.device_info` from where ever you have a reference to the active `protocol` instance. It is still the matter of adding an API for accessing this though. Maybe we can do it like this...

Add a new interface in `interface.py` called `Power` (as I mentioned before) looking something like this:

```python
class PowerListener:
    """Listener interface for power updates."""

    @abstractmethod
    def powerstate_update(self, old_state, new_state):
        """Device power state was updated."""
        raise NotImplementedError()


class Power:
    __metaclass__ = ABCMeta

    def __init__(self):
        """Initialize a new AppleTV."""
        self.__listener = None

    @property
    def listener(self):
        """Object receiving power state updates.

        Must be an object conforming to PowerListener.
        """
        return self.__listener

    @abstractproperty
    def power_state(self):
        """Return device power state."""
        raise exceptions.NotSupportedError()

    @abstractmethod
    def turn_on(self):
        """Turn device on."""
        raise exceptions.NotSupportedError()

    @abstractmethod
    def turn_off(self):
        """Turn device off."""
        raise exceptions.NotSupportedError()

# At the bottom in `AppleTV`, add the following:

    @abstractproperty
    def power(self):
        """Return API for power management."""
        raise exceptions.NotSupportedError()
```

Since we don't yet know how to manage power state with pre-tvOS devices, we need a way to work around that. So my suggestion here is that we also introduce a new state type, which is also more flexible in case we want to add new states in the future. One suggestion is to add the following to `const.py`:

```python
class PowerState(Enum):
    """All supported power states."""

    Unknown = 0
    Off = 1
    On = 2
```

We can hardcode to always return `Unknown` for `DMAP` and let it be either `On` or `Off` for `MRP` depending on value of `logicalDeviceCount`. The `turn_on` and `turn_off` methods can be implemented in any way that fits (basically what you do in the component now). If we find better ways of doing the same thing later, we can improve that.

I also added a listener for power state as well as interface for it (the interface will mostly be relevant when I implement #249). This way we can keep all power related logic in the same place.

Next part is to implement `MrpPower` in `pyatv/mrp/__init__.py` that implements the interface above and satifies the abstract members. This is very similar to how all the interfaces are implemented, `MrpMetadata` can be good to look at as inspiration. I imagine that you can move the message listening logic (`protocol.add_listener` to this class as well). You also need to create an instance of `MrpPower`. This is done in `__init__` in the same file, again you can look at how it's done for `MrpMetadata`.

The same thing also needs to be implemented for `DMAP` (i.e. add `DmapPower` to `pyatv/dmap/__init__.py`, etc). We can work on that over time. We also need to consider tests and documentation in due time.

There's a lot to it here, so I understand if it's a lot to take in. I will help you as much as I can 😄 You basically add completely new API with this, so it pulls a lot of infrastructual code with it. But it's a very good exercise. To not be overwhelmed, do one thing at the time, e.g. focus on `MRP` for now, fix `DMAP`, next, tests after that and finish off with documentation. That makes it a lot easier.

We need to add a few things in `atvrmote` as well, so we can test the feature. Most things are handled automatically, but new interfaces must be added manually at the moment (there's room for improvement here). It follows the same pattern more or less. Do the same thing as for `Metadata` but replace with `Power` in these places:

https://github.com/postlund/pyatv/blob/070946fd05d9c84bbfa2f3f0c98af62a459ec9e3/pyatv/__main__.py#L77
https://github.com/postlund/pyatv/blob/070946fd05d9c84bbfa2f3f0c98af62a459ec9e3/pyatv/__main__.py#L92
https://github.com/postlund/pyatv/blob/070946fd05d9c84bbfa2f3f0c98af62a459ec9e3/pyatv/__main__.py#L493
https://github.com/postlund/pyatv/blob/070946fd05d9c84bbfa2f3f0c98af62a459ec9e3/pyatv/__main__.py#L507

Then you can use `atvremote` and access the API like so:
```shell
$ atvremote ... power_state
$ atvremote ... turn_on
$ atvremote ... turn_off
```